src/storage/src/storage_state.rs

```diff
@@ -230,21 +248,91 @@ impl<'w, A: Allocate> Worker<'w, A> {
 
             self.report_frontier_progress(&response_tx);
 
-            // Handle any received commands.
-            let mut cmds = vec![];
+            // Worker #0 pushes external commands into the sequencer. All other
+            // workers read them away but ignore them. Instead, other workers
+            // will get commands broadcast from the sequencer.
+            //
+            // NOTE: This seems hairy, but I don't yet see an alternative. We
+            // need a consistent order/interleaving of internal and external
+            // commands, across all workers, so we can't let each worker read
+            // their own external command stream because that would be a side
+            // channel and the ordering could (will) be arbitrary.
```

Comment:
I haven't fully grokked what this PR does yet, so this might be entirely misguided, but:

It seems to me that this approach could make the various clients in the communication stack between controller and instance see responses before they see the corresponding commands. For example, consider a replica with two processes:

* Process 0 receives a `CreateSource` command. Process 1 does not receive it yet because the network is slow or whatever.
* Process 0 sends the `CreateSource` command to Process 1 via the sequencer.
* Process 1 creates its half of the source dataflow and continues flowing it.
* Process 1 sends a `FrontierUppers` response to the controller.
* The communication clients living on the instance-side (as opposed to the controller-side) of Process 1 receive the `FrontierUppers` response before they have seen the `CreateSource` response. They might or might not get confused about the update for a frontier they are not tracking.

So it seems to me that even with this change, clients on the instance-side need to expect to not see all commands (in time).

Reply:
yep! You're absolutely right!

I would still have to change the intermediary parts to create the necessary state when they first see a response, similar to how it's done for compute. Overall, I'm somewhat hesitant about the asymmetric messaging approach, but I don't yet see good other ways.

One idea we had initially is for our storage-internal commands to go from worker back to controller, and to then let the controller re-distribute them to all workers. But that seems somewhat complicated, and the idea of an internal control flow looked very appealing.

Reply:
Another approach would be to have all workers funnel frontier updates to worker #0, which coalesces them and forwards them on. You can imagine the controller _only_ connecting out to worker 0, and relying on the control dataflow such that all other communication to and from the controller will flow through that worker.

Has the advantage of shifting some of the burden out of the controller / environmentd. But I have not worked out the details.

Reply:
That drawback of Ben's idea is that it puts more of a burden on worker #0. When we introduced the asymmetric command transport, we were briefly thinking about the load skew but figured it was probably fine. If we introduced asymmetric response transport as well, it might not be fine anymore. Maybe not much of an issue for storage since I think `FrontierUppers` is the only storage response that exists right now. But our goal is to merge compute and storage eventually and compute has `PeekResponse`s and `SubscribeResponse`s that can get quite large.

I wonder if we could shard commands/responses between workers somehow. I.e. each command/response is only received/sent by a single worker, with each worker handling a subset. I guess the difficulty would lie in ensuring that responses flow back through the same worker that received the corresponding command. Otherwise the communication clients would be utterly confused :)

Reply:
Yeah... I feel like it's fine to have a little skew for control-plane messages, but we'd need to be a lot more careful with the data-plane stuff.

Reply:
hrm, it seems Github lost the message I typed up ... 😞 

Taking a step back: maybe it's a problem that `ComputeResponse` is used for _both_ control and data?

For our issue: it'd like to not change how responses are routed back, because it increases the scope of this. Plus we're not 100% whether it's a good solution for the long term.

> Has the advantage of shifting some of the burden out of the controller / environmentd. But I have not worked out the details.

@bkirwi It's already the case that the (storage) controller only has one connection to each storage host, and routing to/from individual workers happens internally. The entrypoint to that is roughly here: https://github.com/MaterializeInc/materialize/blob/6c90a5fad085ce4c359a5c7af7cdddf0d891b372/src/storage/src/server.rs#L133

Reply:
One thing that brings tension here is that what we want to impose a total order on is calls to `Worker::dataflow` which mints ids sequentially but this is done *as part* of `CreateSource` commands which do more than that. If we instead re-interpret these commands to mean "start reporting information for these sources" then we could have all the workers independently receiving them from the storage controller and applying them immediately, but that application would **not** install a dataflow, only start the reporting process.

Now, worker 0 will be special and in addition to reacting to these commands by starting its own reporting process it can additionally publish `CreateDataflow` (different name for clarity) internal command to the sequencer. When the other workers receive a `CreateDataflow` command they will install the dataflow, which is now totally ordered so timely is happy, and start *recording* but not *reporting* their frontiers.

This way we have two independent operations than can be interleaved arbitrarily (both create and drop) and no one gets confused. Reporting starts when each local client stack has seen the `CreateSource` command and dataflows get installed only through internal commands through the `Sequencer`. Frontier recording and frontier reporting eventually rendezvous in storage state using their `GlobalId`.

What do you think?

Reply:
> This way we have two independent operations than can be interleaved arbitrarily (both create and drop) and no one gets confused. Reporting starts when each local client stack has seen the CreateSource command and dataflows get installed only through internal commands through the Sequencer. Frontier recording and frontier reporting eventually rendezvous in storage state using their GlobalId.

I like this idea about turning `CreateDataflows` into an internal command! But I have some questions/comments about the test.

What would "start reporting information" mean? Is it "put the shared frontiers into `storage_state`"? Because other than that, there is nothing we do activate per source/per sink in the worker itself. The reporting machinery is just reporting on all the frontiers that are there. But in that case, we might as well just put those shared frontier bits into `storage_state` when we render the dataflow. In fact, I think we have to do that, because otherwise the shared frontier information is not there if a worker receives the internal `CreateDataflows` command before if sees the controller-issued `CreateSource` command.

If "start reporting information" was meant about the partitioned client (or whatever it's called), then the current PR is actually quite close to that. I still send all the control messages to all workers, only at the very last step are they ignored for workers that are not worker #0. So I could change the partitioned client back to only report information forward once it has seen the relevant `CreateSource` command for that worker.

I'm getting the feeling that maybe we should have a chat on Zoom about this next week. I  _do_ feel we have a pretty good understanding of the design space now, though!

Reply:
> The reporting machinery is just reporting on all the frontiers that are there. But in that case, we might as well just put those shared frontier bits into storage_state when we render the dataflow.

This is exactly what causes the confusion! We shouldn't report on all the frontiers that are there, we should only report the frontiers that we have received a `CreateSource` command for. Concretely, I think storage state needs to change to this:

```rust
struct StorageState {
    recorded_source_uppers: HashMap<GlobalId, Rc<RefCell<Antichain<mz_repr::Timestamp>>>>,
    reported_source_uppers: HashMap<GlobalId, Rc<RefCell<Antichain<mz_repr::Timestamp>>>>,
    ...
}
```

When we receive a `CreateDataflow` we do something like this:

```rust
// Check if the frontier is already being *reported*
let frontier = self.reported_source_uppers[id].cloned().unwrap_or(|| Rc::new(RefCell::new(Antichain::new())));
// Start *recording*
self.recorded_source_uppers.insert(id, frontier);
```

And when we receive a `CreateSource` we do something like this:

```rust
// Check if the frontier is already being *recorded*
let frontier = self.recorded_source_uppers[id].cloned().unwrap_or(|| Rc::new(RefCell::new(Antichain::new())));
// Start *reporting*
self.reported_source_uppers.insert(id, frontier);
```

This way whoever comes first just installs its own private entry. And whoever comes second also installs its own private entry, which shares the frontier via the Rc RefCell.

Agreed on having a high bandwidth zoom on the matter